frontends/ast/simplify.cc

```diff
@@ -2467,18 +2468,28 @@ bool AstNode::simplify(bool const_fold, int stage, int width_hint, bool sign_hin
 			if (member_node)
 				rvalue->set_attribute(ID::wiretype, member_node->clone());
 
-			if (wire_offset != 0) {
-				// Offset the shift amount by the lower bound of the dimension.
-				shift_expr = new AstNode(AST_SUB, shift_expr, node_int(wire_offset));
-			}
-			if (id2ast->range_swapped) {
-				// Reflect the shift amount if the dimension is swapped.
-				shift_expr = new AstNode(AST_SUB, node_int(wire_width - result_width), shift_expr);
-			}
+			// Insert a self-sizing barrier to prevent the rewritten AST from
+			// influencing the size of the operators in `shift_expr`
+			shift_expr = new AstNode(AST_SELFSZ, shift_expr);
+
+			// Decode the index based on wire dimensions
+			int idx_signed_nbits = shift_expr_width_hint + !shift_expr_sign_hint;
+			if (!id2ast->range_swapped) {
+				int raw_idx_nbits = 1 + std::max(idx_signed_nbits, ceil_log2(std::abs(wire_offset) + 1) + 1);
+				shift_expr = new AstNode(AST_SUB,
+					new AstNode(AST_TO_SIGNED,
+						new AstNode(AST_CAST_SIZE, node_int(raw_idx_nbits), shift_expr)
+					),
+				node_int(wire_offset));
```

Comment:
Can you avoid the AST_SUB here if `wire_offset == 0`?
This ends up in RTLIL even when it's not needed.

Reply:
@povik The plot thickens. Skipping the offset calculations as mentioned above, the following code in `tests/simple/shiftx.v` passes `make -f ../tools/autotest.mk shiftx.v`. However if you change the constant from `31'sd0` to `32'sd0`, it fails!

```
module shiftx_test (
	input [1:0] din,
	input [1:0] uoffset,
	output [1:0] dout
);

assign dout = din[31'sd0 -uoffset +: 2];
endmodule
```

Reply:
Holy cow! It seems like iverilog is actually the culprit here! :man_facepalming: 

Simplifying the test even further and running `../tools/autotest.sh -S 1 shiftx.v` yields the same result for `shiftx.out/shiftx_out_syn1` (iverilog via Yosys techmap) for both `31'sd0` and `32'sd0`, while the results in `shiftx.out/shiftx_out_ref` (directly processed by iverilog) are incompatible.

```
module shiftx_test (
	input [1:0] din,
	input uoffset,
	output [1:0] dout
);
	assign dout = din[31'sd0 - uoffset +: 2];
endmodule
```

```
../tools/cmp_tbdata shiftx_out_ref-32 shiftx_out_ref-31
Error in testbench output compare (line=10):
-#OUT# 011 x 1x                 1000           1
+#OUT# 011 x xx                 1000           1
```

Reply:
Here is a (currently not passing) test I cooked up for iverilog, which incidentally only *one* of the commercial simulators on EDA Playground handles correctly :see_no_evil: 
That simulator also warns that it is truncating MSBs if `uoffset` is made wider than 32 bits, which seems reasonable.

```
/*
 * partsel_outside
 * Check that an unsigned integer offset in an indexed part-select is not converted to a signed integer.
 * This would yield incorrect out of bound results for part-selects like arr[idx*8 - uoffset +: 4].
 */

module main;

   reg [1:0] arr = 1;
   int unsigned uoffset = '1;
   wire [1:0] outside = arr[uoffset +: 2];

   initial begin
      #1 if (outside !== 'x) begin
	 $display("FAILED -- out of bounds value %b != xx", outside);
	 $finish;
      end

      $display("PASSED");
      $finish;
   end

endmodule // main
```

Reply:
> Can you avoid the AST_SUB here if wire_offset == 0?
> This ends up in RTLIL even when it's not needed.

We can, but I am convinced to improve the code quality in the frontend, we shouldn't do any specialized optimizations that can be delegated to general machinery -- either to AST transformations (AST_SUB of zero can be removed, if the sizing effect is kept) or netlist passes (`opt_expr` should be able to remove the no-op subtraction later on).

Reply:
> I'm all for improving this code, which I translated more or less directly from genrtlil.cc, however I'd like it to be completely understandable - having to sign an unsigned shift amount doesn't feel quite right 😅

The rationale was the following: We want the result of the subtraction to be signed, since that's the general case, and for that to happen per the Verilog rules, we need both operands signed. Since the signed cast is after resizing, there shouldn't be any hazard of wraparound. Later optimizations can optimize away the sign bit and any extra higher bits, if they are superfluous, so this shouldn't affect QoR negatively.

Reply:
Duly noted, but there is a limit to what spinoff problems I can work on :sweat_smile: 
Currently at least the AST transformations only optimize `AST_ADD`/`AST_SUB` if both operands are constant. In the case of removing addition or subtraction of `0`, one would possibly want to optimize beforehand anyway, since it is not necessary to set aside an extra bit for overflow in this case. One could conceivably introduce new AST nodes for addition and subtraction with self-determined widths guaranteed not to overflow, but I digress.

In any case, testing the simplest possible AST here revealed a bug in iverilog, as far as I can tell. I believe the same kind of bug is actually present in Yosys as well. If I understand correctly, if the index expressions here are >= 32 bits, the additions and subtractions may cause the expressions to wrap around, causing actual vector bits instead of `x` bits to be returned: https://github.com/YosysHQ/yosys/blob/e9cd6ca9e8f9c8c729044e83811762ce13e84a12/frontends/verilog/verilog_parser.y#L955-L965

This may not be a big deal for Yosys, however it's certainly not ideal for a simulator to return anything but `x` bits for out of bounds addresses.

Reply:
> In the case of removing addition or subtraction of 0, one would possibly want to optimize beforehand anyway, since it is not necessary to set aside an extra bit for overflow in this case.

I operate under the assumption that if this bit isn't necessary, it will get optimized away at the netlist stage anyway. If not, I fully agree we need to handle it.

> In any case, testing the simplest possible AST here revealed a bug in iverilog, as far as I can tell.

Feel free to opt for any simple resolution of this PR, since the non-nordshift handling isn't the focus of it. If iverilog doesn't handle something correctly, we can disable that part of the test until it does. The fixed requirement is that the CI passes and that the tests are in a reasonable state, working around shortcomings of external tools should be fine.

Reply:
OK. I've just made some minor simplifications/corrections to your code, keeping the expression widening and addition/subtraction even when it's not strictly needed. This just so happens to pass all iverilog tests both before and after the iverilog PR.

Only the extra widening seems to be left after optimization (not running synthesis), but I guess that doesn't matter much - the widths are already excessive to start with in many cases, since a minimum of 32 bits are used for any offset calculation. In any case excessive widening seems to be pruned after synthesis :+1: